src/main/groovy/org/transmartproject/batch/browsetag/BrowseTagTypeDatabaseReader.groovy

```diff
@@ -0,0 +1,180 @@
+                    bcc.filter_flag,
+                    ati.tag_template_id,
+                    ati.tag_item_id,
+                    ati.tag_item_uid,
+                    ati.display_name,
+                    ati.tag_item_type,
+                    ati.tag_item_subtype,
+                    ati.code_type_name,
+                    ati.required,
+                    att.tag_template_name,
+                    att.tag_template_type,
+                    att.tag_template_subtype
+                FROM $Tables.BIO_CONCEPT_CODE bcc
+                INNER JOIN $Tables.AM_TAG_ITEM ati
+                ON bcc.code_type_name = ati.code_type_name
+                INNER JOIN $Tables.AM_TAG_TEMPLATE att
+                ON ati.tag_template_id = att.tag_template_id
+                WHERE ati.tag_item_id = $tagType.id
+                ORDER BY bcc.code_description ASC
+        """
+    }
+
+    private BrowseFolderType getFolderType(ResultSet rs) {
+        String tagTemplateType = rs.getString('tag_template_type')
+        BrowseFolderType folderType = folderTypes[tagTemplateType]
+        if (folderType == null) {
+            folderType = new BrowseFolderType(
+                    type: tagTemplateType,
+                    displayName: rs.getString('tag_template_name')
+            )
+            folderTypes[tagTemplateType] = folderType
+        }
+        folderType
+    }
+
+    private BrowseTagType getTagType(ResultSet rs) {
+        Long tagItemId = rs.getLong('tag_item_id')
+        BrowseTagType tagType = tagTypes[tagItemId]
+        if (tagType == null) {
+            tagType = new BrowseTagType(
+                    id: tagItemId,
+                    code: rs.getString('code_type_name'),
+                    folderType: getFolderType(rs),
+                    type: rs.getString('tag_item_type'),
+                    subType: rs.getString('tag_item_subtype'),
+                    displayName: rs.getString('display_name'),
+                    required: rs.getBoolean('required')
+            )
+            tagTypes[tagItemId] = tagType
+        }
+        tagType
+    }
+
+    @SuppressWarnings('UnusedPrivateMethodParameter')
+    private BrowseTagType mapRow(ResultSet rs, int rowNum) throws SQLException {
+        def tagType = getTagType(rs)
+        JdbcCursorItemReader<String> valueReader = new JdbcCursorItemReader<>(
+                driverSupportsAbsolute: true,
+                dataSource: dataSource,
+                sql: getValuesSql(tagType),
+                rowMapper: { ResultSet valuesRs, int valuesRowNum ->
+                    valuesRs.getString('code_description')
+                } as RowMapper<String>
+        )
+        ExecutionContext executionContext = new ExecutionContext()
+        valueReader.open(executionContext)
+        def values = []
+        def value
+        while ((value = valueReader.read()) != null) {
+            values.add(value)
+        }
+        valueReader.close()
```

Comment:
The connection and the result set leak if there's an exception during the loop. I also don't get why you're manually managing spring batch resources when you could just have used the (named) jdbcTemplate.